Python bst (#3)
gSchool/ada-CS-Fundamentals-Internship: 03-Binary-Search-Trees/01-Binary-Search-Trees.md

| Binary Tree      | A tree structure in which each node has at most two children, which are referred to as the _left child_ and _right child_. |  | "A Mathematical expression can be written as a binary tree where the leaves are values and the parent nodes are operators like +, -, *, and /." |
| Binary Search Tree      | A specific type of Binary Tree which is a rooted binary tree data structure whose internal nodes each store a key greater than all the keys in the node’s left subtree and less than those in its right subtree. | Ordered Binary Tree | "Because 23 is less than the root, it can only be found in the left subtree of this binary search tree." |
| Leaf      | A node in a tree with no descendants. |  | "Because the left and right subtrees of the current node were `None`, this node is a leaf." |
| Parent      | A node with descendants. |  | "Because the left and right subtrees of the current node not `None`, this node is a parent and not a leaf." |

> | Unbalanced Tree 	|  A BST where each node has 0 or 1 children (it looks like a linked list) 	| | The tree was unbalanced so inserting new values too linear time. |

Reviewer: ```suggestion
| Unbalanced Tree 	|  A tree where most nodes has 0 or 1 children (it looks a lot like a linked list) 	| | The tree was unbalanced so inserting new values took linear time. |
```

| Balanced Tree 	| A tree where the the level of any two leaves differs by at most 1 node.	| | "Because my tree is balanced, it's quick to find values inside it."  |

> | Subtree | The tree which is a child of a node. Note: The name emphasizes that everything which is a descendant of a tree node is a tree, too, and is a subset of the larger tree.| | "The value I'm looking for is less than the current node, so I will continue searching in the left subtree." |

Reviewer: Things like this are why I like having the `Node` class just be the `Tree` class.  It really makes it clear subtrees are also trees.

> | Traversal 	|  A method of visiting each node in a BST	| | "My method performs a traversal visiting each node in the tree, level-by-level." |

Reviewer: Consider changing to "data structure" to plant seeds for graphs:

```suggestion
| Traversal 	|  A method of visiting each node in a data structure	| | "My method performs a traversal visiting each node in the tree, level-by-level." |
```

> | Depth-First Traversal 	|  An algorithm for traversing or searching a tree. The algorithm starts at the root node and explores as far as possible along each branch before backtracking.	| | A depth-first traversal looks like a mouse exploring a maze in that it goes as far down one path before backtracking when encountering dead ends. |

Reviewer: ```suggestion
| Depth-First Traversal 	|  An algorithm for traversing or searching a data structure. The algorithm starts at the root node and explores as far as possible along each branch before backtracking.	| | A depth-first traversal looks like a mouse exploring a maze in that it goes as far down one path before backtracking when encountering dead ends. |
```

> | Breadth-first traversal 	|  An algorithm for searching a tree data structure for a node that satisfies a given property. It starts at the tree root and explores all nodes at the present depth prior to moving on to the nodes at the next depth level. 	| | I printed out the tree level-by-level so I had to perform a breadth-first traversal. |

Reviewer: ```suggestion
| Breadth-first traversal 	|  An algorithm for searching a data structure for a node that satisfies a given property. It starts at the tree root and explores all nodes at the present depth prior to moving on to the nodes at the next depth level. 	| | I printed out the tree level-by-level so I had to perform a breadth-first traversal. |
```

In a Binary Search Tree:
- Nodes with values less than any node are stored to the **left** of that node.
- Nodes with values greater than any node are stored to the **right** of that node. 

![Binary Search Tree Vocabular](images/TreeVocabulary.png)

### Binary Search Tree Node

Instead of a node with one link to a next node, we can create nodes with 2 pointers, left and right.  Since each node can have 2 successors or "children", it forms a binary structure as opposed to the linear structure of a Linked List.

This node stores both a key and value for each node.  The `Tree` class will compare keys to maintain node order.

```python
class TreeNode:
    def __init__(self, key, val = None):
        if val == None:

>             val = key

Reviewer: Is this to fix the bug with the tests?

        self.key = key
        self.value = val
        self.left = None

>         self.right = None```

Reviewer: The diff makes it look like you duplicated ` ``` `s:

```suggestion
        self.right = None
```

Notice if a tree is balanced, when you move left or right, you eliminate half of the possible nodes.  This means you are essentially doing **binary search.**  If the tree is unbalanced, you are performing a linear search.

<!-- available callout types: info, success, warning, danger, secondary, star  -->
### !callout-star

## Balance is important in a tree's efficiency

If a tree is _balanced_ then adding, finding, removing operations on a that tree perform in O(log n) time.  However if a tree becomes unbalanced the efficiency can approach O(n) time complexity.  

> For this reason, computer scientists spend a lot of time focusing on ways to maintain the balance of a Binary Search Tree.

Reviewer: I really like this call out. :)
